File: interop/definitions.py
"""Game, loader, and runtime definition parsing for the vector corpus."""

from fields import (
    Fields,
    array,
    artifact_ref,
    blob,
    boolean,
    expect_type,
    i64,
    one_of,
    predicate,
    text,
    text_array,
    u32,
)
from reject import Reject

ORDERING_SCHEMES = {"semver", "ordered-list", "calendar", "opaque"}
RUNTIME_KINDS = {"java", "dotnet", "node", "native", "other"}
QUALIFICATIONS = {"native", "most", "experimental", "untested"}


def version_syntax(value):
    fields = Fields("VersionSyntax", value).known(["kind", "pattern"])
    one_of(fields.required("kind"), "kind", ORDERING_SCHEMES)
    if fields.optional("pattern") is not None:
        text(fields.optional("pattern"), "pattern")


def definition_category(value):
    fields = Fields("Category", value).known(["id", "label", "parent"])
    if text(fields.required("id"), "id") == "":
        raise Reject("invalid-field-value", "id")
    text(fields.required("label"), "label")
    if fields.optional("parent") is not None:
        text(fields.optional("parent"), "parent")
    return fields.value.get("id")


def definition_tag(value):
    fields = Fields("Tag", value).known(["id", "label"])
    if text(fields.required("id"), "id") == "":
        raise Reject("invalid-field-value", "id")
    text(fields.required("label"), "label")
    return fields.value.get("id")


def reject_duplicates(ids, key):
    seen = set()
    for item in ids:
        if item in seen:
            raise Reject("invalid-field-value", key)
        seen.add(item)


def version_catalog(value, name="version_catalog"):
    if value is None:
        return []
    entries = text_array(value, name)
    if any(entry == "" for entry in entries):
        raise Reject("invalid-field-value", name)
    reject_duplicates(entries, name)
    return entries


def parse_game_def(value):
    fields = Fields("GameDef", value).known(
        [
            "protocol",
            "game_id",
            "display_name",
            "version_syntax",
            "version_ordering",
            "version_catalog",
            "loaders_allowed",
            "loader_authorities",
            "categories",
            "tags",
            "metadata_extractor",
            "install_adapter",
            "declared_time",
        ]
    )
    protocol = u32(fields.required("protocol"), "protocol")
    game_id = text(fields.required("game_id"), "game_id")
    display_name = text(fields.required("display_name"), "display_name")
    version_syntax(fields.required("version_syntax"))
    ordering = one_of(
        fields.required("version_ordering"), "version_ordering", ORDERING_SCHEMES
    )
    version_catalog(fields.optional("version_catalog"))
    boolean(fields.required("loaders_allowed"), "loaders_allowed")
    text_array(fields.required("loader_authorities"), "loader_authorities")
    categories = [
        definition_category(item)
        for item in array(fields.required("categories"), "categories")
    ]
    tags = [definition_tag(item) for item in array(fields.required("tags"), "tags")]
    if fields.optional("metadata_extractor") is not None:
        text(fields.optional("metadata_extractor"), "metadata_extractor")
    if fields.optional("install_adapter") is not None:
        text(fields.optional("install_adapter"), "install_adapter")
    i64(fields.required("declared_time"), "declared_time")
    if protocol != 1:
        raise Reject("invalid-field-value", "protocol")
    if game_id == "" or display_name == "":
        raise Reject("invalid-field-value", "game_id")
    if ordering is None:
        raise Reject("invalid-field-value", "version_ordering")
    reject_duplicates(categories, "categories")
    reject_duplicates(tags, "tags")


def parse_runtime_def(value):
    fields = Fields("RuntimeDef", value).known(
        [
            "protocol",
            "runtime_id",
            "kind",
            "display_name",
            "version_ordering",
            "version_catalog",
            "declared_time",
        ]
    )
    protocol = u32(fields.required("protocol"), "protocol")
    runtime_id = text(fields.required("runtime_id"), "runtime_id")
    one_of(fields.required("kind"), "kind", RUNTIME_KINDS)
    text(fields.required("display_name"), "display_name")
    one_of(fields.required("version_ordering"), "version_ordering", ORDERING_SCHEMES)
    version_catalog(fields.optional("version_catalog"))
    i64(fields.required("declared_time"), "declared_time")
    if protocol != 1:
        raise Reject("invalid-field-value", "protocol")
    if runtime_id == "":
        raise Reject("invalid-field-value", "runtime_id")


def declared_by(value):
    fields = Fields("DeclaredBy", value).known(["kind", "id"])
    one_of(fields.required("kind"), "kind", {"loader-authority", "project", "tester"})
    if text(fields.required("id"), "id") == "":
        raise Reject("invalid-field-value", "kind")


def parse_loader_object(value):
    from cbor import Map

    if not isinstance(value, Map):
        raise Reject("invalid-field-type", "LoaderObject")
    discriminant = value.get("type")
    if type(discriminant) is not str:
        raise Reject("missing-field", "type")
    if discriminant == "definition":
        loader_definition(value)
        return
    if discriminant == "release":
        loader_release(value)
        return
    if discriminant == "mapping":
        loader_acceptance(value)
        return
    raise Reject("invalid-field-value", "type")


def loader_definition(value):
    fields = Fields("LoaderDef", value).known(
        [
            "protocol",
            "type",
            "loader_id",
            "game_id",
            "display_name",
            "version_ordering",
            "version_catalog",
            "game_versions",
            "bootstrap",
            "accepted_artifacts",
            "declared_time",
        ]
    )
    expect_type(fields, "definition")
    protocol = u32(fields.required("protocol"), "protocol")
    loader_id = text(fields.required("loader_id"), "loader_id")
    text(fields.required("game_id"), "game_id")
    text(fields.required("display_name"), "display_name")
    ordering = one_of(
        fields.required("version_ordering"), "version_ordering", ORDERING_SCHEMES
    )
    version_catalog(fields.optional("version_catalog"))
    if fields.optional("game_versions") is not None:
        predicate(fields.optional("game_versions"))
    if fields.optional("bootstrap") is not None:
        artifact_ref(fields.optional("bootstrap"))
    if fields.optional("accepted_artifacts") is not None:
        text_array(fields.optional("accepted_artifacts"), "accepted_artifacts")
    i64(fields.required("declared_time"), "declared_time")
    if protocol != 1:
        raise Reject("invalid-field-value", "protocol")
    if loader_id == "":
        raise Reject("invalid-field-value", "loader_id")
    if ordering is None:
        raise Reject("invalid-field-value", "version_ordering")


def loader_release(value):
    fields = Fields("LoaderRelease", value).known(
        [
            "protocol",
            "type",
            "loader_id",
            "version_id",
            "game_version_predicate",
            "runtime_id",
            "runtime_predicate",
            "bootstrap",
            "declared_time",
        ]
    )
    expect_type(fields, "release")
    protocol = u32(fields.required("protocol"), "protocol")
    loader_id = text(fields.required("loader_id"), "loader_id")
    version_id = text(fields.required("version_id"), "version_id")
    predicate(fields.required("game_version_predicate"))
    if fields.optional("runtime_id") is not None:
        text(fields.optional("runtime_id"), "runtime_id")
    if fields.optional("runtime_predicate") is not None:
        predicate(fields.optional("runtime_predicate"))
    if fields.optional("bootstrap") is not None:
        artifact_ref(fields.optional("bootstrap"))
    i64(fields.required("declared_time"), "declared_time")
    if protocol != 1:
        raise Reject("invalid-field-value", "protocol")
    if loader_id == "" or version_id == "":
        raise Reject("invalid-field-value", "version_id")


def loader_acceptance(value):
    fields = Fields("LoaderAcceptance", value).known(
        [
            "protocol",
            "type",
            "accepting_loader_id",
            "accepted_loader_id",
            "game_id",
            "game_version_predicate",
            "loader_version_predicate",
            "accepted_version_predicate",
            "qualification",
            "declared_by",
            "evidence_digest",
            "declared_time",
        ]
    )
    expect_type(fields, "mapping")
    protocol = u32(fields.required("protocol"), "protocol")
    accepting = text(fields.required("accepting_loader_id"), "accepting_loader_id")
    accepted = text(fields.required("accepted_loader_id"), "accepted_loader_id")
    text(fields.required("game_id"), "game_id")
    for name in (
        "game_version_predicate",
        "loader_version_predicate",
        "accepted_version_predicate",
    ):
        if fields.optional(name) is not None:
            predicate(fields.optional(name))
    one_of(fields.required("qualification"), "qualification", QUALIFICATIONS)
    declared_by(fields.required("declared_by"))
    if fields.optional("evidence_digest") is not None:
        raw = blob(fields.optional("evidence_digest"), "evidence_digest")
        if len(raw) != 32:
            raise Reject("invalid-field-value", "evidence_digest")
    i64(fields.required("declared_time"), "declared_time")
    if protocol != 1:
        raise Reject("invalid-field-value", "protocol")
    if accepting == "" or accepted == "":
        raise Reject("invalid-field-value", "accepting_loader_id")

File: verify/src/definitions.rs
use moraine_codec::Value;
use moraine_crypto::ObjectKind;
use moraine_model::Canonical;
use moraine_model::compatibility::{Predicate, Scheme};
use moraine_model::definition::{
	Category, DeclaredBy, GameDef, LoaderAcceptance, LoaderDef, LoaderObject, LoaderRelease, Qualification, RuntimeDef, Tag,
	VersionSyntax,
};
use moraine_model::reference::ArtifactRef;
use moraine_model::signed::sign_payload;
use moraine_model::version::{OrderingScheme, VersionCatalog};

use crate::generate::{
	DECLARED_AT, Outcome, object_vector, predicate_vector, public_hex, raw_vector, sample_id, signer, trust,
};
use crate::vector::Vector;

fn pv(name: &str, ordering: &str, catalog: &[&str], scheme: &str, values: &[&str], version: &str, expected: &str) -> Vector {
	let ordering = OrderingScheme::parse(ordering).expect("known ordering scheme");
	let catalog = VersionCatalog::new(ordering, catalog.iter().map(|value| value.to_string()).collect());
	let predicate = Predicate {
		scheme: scheme.to_string(),
		values: values.iter().map(|value| value.to_string()).collect(),
	};
	predicate_vector(name, catalog, predicate, version, expected)
}

fn category(id: &str) -> Category {
	Category {
		id: id.to_string(),
		label: id.to_string(),
		parent: None,
	}
}

fn tag(id: &str) -> Tag {
	Tag {
		id: id.to_string(),
		label: id.to_string(),
	}
}

fn build_game_def(version_ordering: &str, categories: Vec<Category>) -> GameDef {
	GameDef {
		protocol: 1,
		game_id: sample_id("minecraft"),
		display_name: "Minecraft".to_string(),
		version_syntax: VersionSyntax {
			kind: "semver".to_string(),
			pattern: None,
		},
		version_ordering: version_ordering.to_string(),
		version_catalog: Vec::new(),
		loaders_allowed: true,
		loader_authorities: vec![sample_id("fabric-authority")],
		categories,
		tags: vec![tag("client"), tag("server")],
		metadata_extractor: Some("minecraft/fabric-json".to_string()),
		install_adapter: Some("minecraft/default".to_string()),
		declared_time: DECLARED_AT,
	}
}

fn build_runtime_def() -> RuntimeDef {
	RuntimeDef {
		protocol: 1,
		runtime_id: sample_id("java"),
		kind: "java".to_string(),
		display_name: "Java".to_string(),
		version_ordering: "semver".to_string(),
		version_catalog: Vec::new(),
		declared_time: DECLARED_AT,
	}
}

fn build_loader_def() -> LoaderDef {
	LoaderDef {
		protocol: 1,
		loader_id: sample_id("fabric"),
		game_id: sample_id("minecraft"),
		display_name: "Fabric".to_string(),
		version_ordering: "semver".to_string(),
		version_catalog: Vec::new(),
		game_versions: None,
		bootstrap: Some(ArtifactRef {
			digest: vec![0x55; 32],
			size: 4096,
			media_type: "application/java-archive".to_string(),
		}),
		accepted_artifacts: None,
		declared_time: DECLARED_AT,
	}
}

fn build_loader_release() -> LoaderRelease {
	LoaderRelease {
		protocol: 1,
		loader_id: sample_id("fabric"),
		version_id: "0.15.0".to_string(),
		game_version_predicate: Predicate::new(Scheme::Exact, vec!["1.20.1".to_string()]),
		runtime_id: Some(sample_id("java")),
		runtime_predicate: None,
		bootstrap: None,
		declared_time: DECLARED_AT,
	}
}

fn build_loader_acceptance() -> LoaderAcceptance {
	LoaderAcceptance {
		protocol: 1,
		accepting_loader_id: sample_id("cleanroom"),
		accepted_loader_id: sample_id("forge"),
		game_id: sample_id("minecraft"),
		game_version_predicate: Some(Predicate::new(Scheme::Exact, vec!["1.12.2".to_string()])),
		loader_version_predicate: None,
		accepted_version_predicate: None,
		qualification: Qualification::Most,
		declared_by: DeclaredBy {
			kind: "loader-authority".to_string(),
			id: sample_id("cleanroom-authority"),
		},
		evidence_digest: None,
		declared_time: DECLARED_AT,
	}
}

fn acceptance_value(qualification: &str) -> Value {
	Value::map([
		(Value::text("protocol"), Value::int(1)),
		(Value::text("type"), Value::text("mapping")),
		(Value::text("accepting_loader_id"), Value::text(sample_id("cleanroom"))),
		(Value::text("accepted_loader_id"), Value::text(sample_id("forge"))),
		(Value::text("game_id"), Value::text(sample_id("minecraft"))),
		(Value::text("qualification"), Value::text(qualification)),
		(
			Value::text("declared_by"),
			Value::map([
				(Value::text("kind"), Value::text("loader-authority")),
				(Value::text("id"), Value::text(sample_id("cleanroom-authority"))),
			]),
		),
		(Value::text("declared_time"), Value::int(DECLARED_AT)),
	])
}

pub(crate) fn vectors() -> Result<Vec<Vector>, String> {
	let k1 = signer(1);
	let k2 = signer(2);
	let mut vectors = Vec::with_capacity(64);

	vectors.push(pv(
		"predicate-semver-prefix",
		"semver",
		&[],
		"semver",
		&["1.20"],
		"1.20.4",
		"satisfied",
	));
	vectors.push(pv(
		"predicate-semver-prefix-miss",
		"semver",
		&[],
		"semver",
		&["1.20"],
		"1.21.0",
		"not-satisfied",
	));
	vectors.push(pv(
		"predicate-semver-range",
		"semver",
		&[],
		"semver",
		&[">=1.4.0", "<2.0.0"],
		"1.9.9",
		"satisfied",
	));
	vectors.push(pv(
		"predicate-semver-range-exceeds",
		"semver",
		&[],
		"semver",
		&[">=1.4.0", "<2.0.0"],
		"2.0.0",
		"not-satisfied",
	));
	vectors.push(pv(
		"predicate-semver-prerelease",
		"semver",
		&[],
		"semver",
		&["=1.0.0-alpha"],
		"1.0.0-alpha",
		"satisfied",
	));
	vectors.push(pv(
		"predicate-semver-prerelease-order",
		"semver",
		&[],
		"semver",
		&["=1.0.0-alpha"],
		"1.0.0",
		"not-satisfied",
	));
	vectors.push(pv(
		"predicate-exact-hit",
		"semver",
		&[],
		"exact",
		&["1.20.1"],
		"1.20.1",
		"satisfied",
	));
	vectors.push(pv(
		"predicate-exact-miss",
		"semver",
		&[],
		"exact",
		&["1.20.1"],
		"1.20.2",
		"not-satisfied",
	));
	vectors.push(pv("predicate-any", "semver", &[], "any", &[], "9.9.9", "satisfied"));
	vectors.push(pv(
		"predicate-unknown-scheme",
		"semver",
		&[],
		"maven",
		&["[1.0,2.0)"],
		"1.5",
		"unknown",
	));
	vectors.push(pv(
		"predicate-scheme-mismatch",
		"semver",
		&[],
		"ordered-list",
		&[],
		"1.5",
		"unknown",
	));

	let catalog = ["1.16", "1.17", "1.18", "1.19", "1.20", "1.20.4"];
	vectors.push(pv(
		"predicate-ordered-range-exclusive",
		"ordered-list",
		&catalog,
		"ordered-list",
		&["1.17..1.20.4"],
		"1.19",
		"satisfied",
	));
	vectors.push(pv(
		"predicate-ordered-range-excludes-end",
		"ordered-list",
		&catalog,
		"ordered-list",
		&["1.17..1.20.4"],
		"1.20.4",
		"not-satisfied",
	));
	vectors.push(pv(
		"predicate-ordered-range-inclusive",
		"ordered-list",
		&catalog,
		"ordered-list",
		&["1.17..=1.20.4"],
		"1.20.4",
		"satisfied",
	));
	vectors.push(pv(
		"predicate-ordered-unknown-version",
		"ordered-list",
		&catalog,
		"ordered-list",
		&["1.17..1.20.4"],
		"1.21",
		"unknown",
	));

	let calendar = ["2024-01", "2024-02", "2024-03"];
	vectors.push(pv(
		"predicate-calendar-range",
		"calendar",
		&calendar,
		"calendar",
		&["2024-01..=2024-03"],
		"2024-02",
		"satisfied",
	));
	vectors.push(pv(
		"predicate-calendar-outside",
		"calendar",
		&calendar,
		"calendar",
		&["2024-01..2024-02"],
		"2024-03",
		"not-satisfied",
	));

	let game = build_game_def("semver", vec![category("utility"), category("qol")]);
	let signed_game = sign_payload(ObjectKind::GameDef, &game, &[&k1]).map_err(|error| error.to_string())?;
	vectors.push(object_vector(
		"game-def-valid",
		"definitions",
		ObjectKind::GameDef,
		&signed_game,
		"accept",
		None,
		Some(trust(&[&k1], 1)),
	));

	let bad_ordering = build_game_def("lexicographic", vec![category("utility")]);
	vectors.push(raw_vector(
		"game-def-unknown-ordering",
		"definitions",
		ObjectKind::GameDef,
		bad_ordering.to_value(),
		&[&k1],
		Outcome {
			verdict: "reject",
			reason: Some("invalid-field-value"),
		},
		Some(trust(&[&k1], 1)),
	));

	let duplicate = build_game_def("semver", vec![category("utility"), category("utility")]);
	vectors.push(raw_vector(
		"game-def-duplicate-category",
		"definitions",
		ObjectKind::GameDef,
		duplicate.to_value(),
		&[&k1],
		Outcome {
			verdict: "reject",
			reason: Some("invalid-field-value"),
		},
		Some(trust(&[&k1], 1)),
	));

	let runtime = build_runtime_def();
	let signed_runtime = sign_payload(ObjectKind::RuntimeDef, &runtime, &[&k1]).map_err(|error| error.to_string())?;
	vectors.push(object_vector(
		"runtime-def-valid",
		"definitions",
		ObjectKind::RuntimeDef,
		&signed_runtime,
		"accept",
		None,
		Some(trust(&[&k1], 1)),
	));

	let mut catalog_game = build_game_def("ordered-list", vec![category("utility")]);
	catalog_game.version_catalog = vec!["1.0".to_string(), "1.1".to_string(), "1.2".to_string()];
	let signed_catalog_game = sign_payload(ObjectKind::GameDef, &catalog_game, &[&k1]).map_err(|error| error.to_string())?;
	vectors.push(object_vector(
		"game-def-version-catalog",
		"definitions",
		ObjectKind::GameDef,
		&signed_catalog_game,
		"accept",
		None,
		Some(trust(&[&k1], 1)),
	));

	let mut duplicate_catalog = build_game_def("ordered-list", vec![category("utility")]);
	duplicate_catalog.version_catalog = vec!["1.0".to_string(), "1.0".to_string()];
	vectors.push(raw_vector(
		"game-def-duplicate-version-catalog",
		"definitions",
		ObjectKind::GameDef,
		duplicate_catalog.to_value(),
		&[&k1],
		Outcome {
			verdict: "reject",
			reason: Some("invalid-field-value"),
		},
		Some(trust(&[&k1], 1)),
	));

	let mut empty_catalog = build_game_def("ordered-list", vec![category("utility")]);
	empty_catalog.version_catalog = vec![String::new()];
	vectors.push(raw_vector(
		"game-def-empty-version-catalog-entry",
		"definitions",
		ObjectKind::GameDef,
		empty_catalog.to_value(),
		&[&k1],
		Outcome {
			verdict: "reject",
			reason: Some("invalid-field-value"),
		},
		Some(trust(&[&k1], 1)),
	));

	let mut catalog_runtime = build_runtime_def();
	catalog_runtime.version_catalog = vec!["17".to_string(), "21".to_string()];
	let signed_catalog_runtime =
		sign_payload(ObjectKind::RuntimeDef, &catalog_runtime, &[&k1]).map_err(|error| error.to_string())?;
	vectors.push(object_vector(
		"runtime-def-version-catalog",
		"definitions",
		ObjectKind::RuntimeDef,
		&signed_catalog_runtime,
		"accept",
		None,
		Some(trust(&[&k1], 1)),
	));

	let loader = build_loader_def();
	let signed_loader =
		sign_payload(ObjectKind::LoaderDef, &LoaderObject::Definition(loader), &[&k1]).map_err(|error| error.to_string())?;
	vectors.push(object_vector(
		"loader-def-valid",
		"definitions",
		ObjectKind::LoaderDef,
		&signed_loader,
		"accept",
		None,
		Some(trust(&[&k1], 1)),
	));

	let mut catalog_loader = build_loader_def();
	catalog_loader.version_catalog = vec!["0.15.0".to_string(), "0.16.0".to_string()];
	catalog_loader.game_versions = Some(Predicate::new(Scheme::Semver, vec![">=1.20.1".to_string()]));
	let signed_catalog_loader = sign_payload(ObjectKind::LoaderDef, &LoaderObject::Definition(catalog_loader), &[&k1])
		.map_err(|error| error.to_string())?;
	vectors.push(object_vector(
		"loader-def-version-catalog-and-game-versions",
		"definitions",
		ObjectKind::LoaderDef,
		&signed_catalog_loader,
		"accept",
		None,
		Some(trust(&[&k1], 1)),
	));

	let mut duplicate_loader_catalog = build_loader_def();
	duplicate_loader_catalog.version_catalog = vec!["0.15.0".to_string(), "0.15.0".to_string()];
	vectors.push(raw_vector(
		"loader-def-duplicate-version-catalog",
		"definitions",
		ObjectKind::LoaderDef,
		LoaderObject::Definition(duplicate_loader_catalog).to_value(),
		&[&k1],
		Outcome {
			verdict: "reject",
			reason: Some("invalid-field-value"),
		},
		Some(trust(&[&k1], 1)),
	));

	let release = build_loader_release();
	let signed_release =
		sign_payload(ObjectKind::LoaderDef, &LoaderObject::Release(release), &[&k1]).map_err(|error| error.to_string())?;
	vectors.push(object_vector(
		"loader-release-valid",
		"definitions",
		ObjectKind::LoaderDef,
		&signed_release,
		"accept",
		None,
		Some(trust(&[&k1], 1)),
	));

	let acceptance = build_loader_acceptance();
	let signed_acceptance = sign_payload(ObjectKind::LoaderDef, &LoaderObject::Acceptance(acceptance), &[&k1])
		.map_err(|error| error.to_string())?;
	vectors.push(object_vector(
		"loader-acceptance-directional",
		"loader-acceptance",
		ObjectKind::LoaderDef,
		&signed_acceptance,
		"accept",
		None,
		Some(trust(&[&k1], 1)),
	));

	let invalid_qualification = acceptance_value("definitely-not-valid");
	vectors.push(raw_vector(
		"loader-acceptance-invalid-qualification",
		"loader-acceptance",
		ObjectKind::LoaderDef,
		invalid_qualification,
		&[&k1],
		Outcome {
			verdict: "reject",
			reason: Some("invalid-field-value"),
		},
		Some(trust(&[&k1], 1)),
	));

	let mut extra_field = acceptance_value("most");
	if let Value::Map(pairs) = &mut extra_field {
		pairs.push((Value::text("oops"), Value::int(1)));
	}
	vectors.push(raw_vector(
		"loader-acceptance-unknown-field",
		"loader-acceptance",
		ObjectKind::LoaderDef,
		extra_field,
		&[&k1],
		Outcome {
			verdict: "reject",
			reason: Some("unknown-field"),
		},
		Some(trust(&[&k1], 1)),
	));

	let missing_field = Value::map([
		(Value::text("protocol"), Value::int(1)),
		(Value::text("type"), Value::text("release")),
		(Value::text("loader_id"), Value::text(sample_id("fabric"))),
		(
			Value::text("game_version_predicate"),
			Value::map([
				(Value::text("scheme"), Value::text("exact")),
				(Value::text("values"), Value::array([Value::text("1.20.1")])),
			]),
		),
		(Value::text("declared_time"), Value::int(DECLARED_AT)),
	]);
	vectors.push(raw_vector(
		"loader-release-missing-version",
		"definitions",
		ObjectKind::LoaderDef,
		missing_field,
		&[&k1],
		Outcome {
			verdict: "reject",
			reason: Some("missing-field"),
		},
		Some(trust(&[&k1], 1)),
	));

	let signed_by_delegated = sign_payload(ObjectKind::GameDef, &game, &[&k2]).map_err(|error| error.to_string())?;
	let mut delegated_trust = trust(&[&k1], 1);
	delegated_trust.delegated_keys = vec![public_hex(&k2)];
	vectors.push(object_vector(
		"game-def-bad-signature",
		"definitions",
		ObjectKind::GameDef,
		&signed_by_delegated,
		"reject",
		Some("bad-signature"),
		Some(delegated_trust),
	));

	Ok(vectors)
}
